File: kotlin-node/karakum/src/jsMain/resources/plugins/convertStreamOptionsTypeParameter.js
import ts from "typescript";

export default function (node, context, render) {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"
    if (!sourceFileName.endsWith("stream.d.ts")) return null

    if (
        ts.isIdentifier(node)
        && (
            node.text === "Readable"
            || node.text === "Writable"
        )

        && node?.parent?.parent // Identifier / TypeReference / ExpressionWithTypeArguments
        && ts.isExpressionWithTypeArguments(node.parent.parent)
        && ts.isIdentifier(node.parent.parent.expression)
        && node.parent.parent.expression.text === "StreamOptions"

        && node?.parent?.parent?.parent?.parent // Identifier / TypeReference / ExpressionWithTypeArguments / HeritageClause / InterfaceDeclaration
        && ts.isInterfaceDeclaration(node.parent.parent.parent.parent)
        && (
            node.parent.parent.parent.parent.name.text === "ReadableOptions"
            || node.parent.parent.parent.parent.name.text === "WritableOptions"
        )
    ) {
        return `Stream /* ${render(node)} */`
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertStreamOptionsTypeParameter.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.util.getParentOrNull
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import typescript.isExpressionWithTypeArguments
import typescript.isIdentifier
import typescript.isInterfaceDeclaration

val convertStreamOptionsTypeParameter = createPlugin { node, _, render ->
    nullable {
        val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName
        ensure(sourceFileName.endsWith("stream.d.ts"))

        ensure(isIdentifier(node))
        ensure(
            node.text == "Readable"
                    || node.text == "Writable"
        )

        val expressionWithTypeArguments = node
            .getParentOrNull() // TypeReference
            ?.getParentOrNull() // ExpressionWithTypeArguments
            .let { ensureNotNull(it) }
        ensure(isExpressionWithTypeArguments(expressionWithTypeArguments))

        val expression = expressionWithTypeArguments.expression
        ensure(isIdentifier(expression))
        ensure(expression.text == "StreamOptions")

        val interfaceNode = expressionWithTypeArguments
            .getParentOrNull() // HeritageClause
            ?.getParentOrNull() // InterfaceDeclaration
            .let { ensureNotNull(it) }
        ensure(isInterfaceDeclaration(interfaceNode))
        ensure(
            interfaceNode.name.text == "ReadableOptions"
                    || interfaceNode.name.text == "WritableOptions"
        )

        "Stream /* ${render(node)} */"
    }
}
